core/shared/src/main/scala/zio/ZIO.scala

```diff
@@ -2232,14 +2232,8 @@ object ZIO {
    * For a sequential version of this method, see `foreach_`.
    */
   def foreachPar_[R, E, A](as: Iterable[A])(f: A => ZIO[R, E, Any]): ZIO[R, E, Unit] =
-    ZIO
-      .effectTotal(as.iterator)
-      .flatMap { i =>
-        def loop(a: A): ZIO[R, E, Unit] =
-          if (i.hasNext) f(a).zipWithPar(loop(i.next))((_, _) => ())
-          else f(a).unit
-        if (i.hasNext) loop(i.next)
-        else ZIO.unit
+    as.foldLeft(unit: ZIO[R, E, Unit]) { (acc, a) =>
```

Comment:
It looks like you could avoid the `foldLeft` just by making `loop` tail recursive:
```scala
@tailrec def loop(a: A, acc: ZIO[R, E, Any]): ZIO[R, E, Any] =
  if (i.hasNext)
    loop(i.next, acc <&> f(a))
  else
    acc

if (i.hasNext) loop(i.next, ZIO.unit).unit else ZIO.unit
```

Reply:
woops, small mistake, I corrected.

Reply:
In my opinion, version with `foldLeft` is a lot simpler. Also, consiming one more stack frame doesn't seem like a big deal.

Why should we avoid `foldLeft` here?

Reply:
I may be wrong but `foldLeft` will box primitive types no matter what, whereas using a while loop won't.

Reply:
Usually, they will be already boxed anyway inside `as: Iterable[A]`. So, `foldLeft` shouldn't do any boxing again in that case, at the first glance.